# TopONE (Topological Omics for Non-tree Evolution)

This project aims to understand how recombination events influence various topological quantities of a genomic dataset using topological data analysis (TDA). The project seeks to determine whether topology can detect recombination events. More specifically, this project asks the following questions :

1. Is topology robust to noise and sparse genomic samples?
2. Does recombination change the topology of genomic samples?

The project's repository makes use of scripts and sequence files to answer the questions above. This project was made during the 2024 Bioinformatics Training and Internship Program headed by the Philippine Genome Center.

## Installation
### Population Genetics Software `ms`
`ms` is a population genetics software that generates samples under different models. The source code and software manual containing the installation instructions are found [here](https://uchicago.app.box.com/s/l3e5uf13tikfjm7e1il1eujitlsjdx13).

### Sequence Alignment Software `nextclade`
Nextclade is an online [web application](https://clades.nextstrain.org/) bioinformatics tool to align sequences and conduct phylogenetic analyses. It also has a CLI version found [here](https://docs.nextstrain.org/projects/nextclade/en/stable/user/nextclade-cli/installation/standalone.html), which performs the same functionalities as the web application version. Download the the latest version of the software based on your architecture (i.e. x86-64) and operating system.

### Required Libraries
The list of libraries used by this project are listed in the `requirements.txt` file. To install the libraries, use the command:

`pip install -r requirements.txt`

to install all the libraries at the specified versions.

## Pipelines
### Goal 01: Is topology robust to noise and sparse genomic samples?
<a href="url"><img src="https://github.com/alfonsomiguelcruz/pgc_btip_topone/blob/main/imgs/goal01_pipeline.png" height="50%" width="50%" ></a>

The pipeline begins with generating the simulated sequences from `ms` with defined parameters: number of samples, number of simulations, number of recombination sites, number of segregation sites, and the mutation and recombination rates. The output is a group of sequences in each simulation stored as a text file, where all sequences are in a biallelic format, which contains `0` or `1` as its only characters in each sequence. These sequences are then extracted from the text files. A sample of the generated sequences are located in the [simseq](inputs/simseq/) folder.

Once the sequences have been produced, the Hamming distance matrix is generated by computing the Hamming distances of all sequences against each other, for every simulation.

After creating the Hamming distance matrix, it is passed to [Ripser](https://ripser.scikit-tda.org/en/latest/), a persistent homology library, to compute for the birth and death times of each homology group from $H_0$ to $H_2$. This process is repeated for every simulation.

Once the birth and death time pairs have been computed, the Betti numbers from $\beta_0$ to $\beta_2$, barcode length means, and barcode length variance, collectively known as the topological quantities, are computed for each simulation. These values are used for plotting and statistical analysis.

To investigate on how sparsity affects the topology, a fraction of the sequences from the population are taken before computing the Hamming distance matrices. We take a fraction of the population in values from 10% to 100% in increasing 10% increments to see any changes in the values of the topological quantities.

To investigate on how noise affects the topology and its quantities, we add a noise matrix to the computed Hamming distance matrix. This is done by creating a matrix $N$ with the same dimensions as the Hamming distance matrix. Each element $N_{ij}$ is sampled from a normal distribution $\mathcal{N}(0, \sigma^2)$, where $\sigma^2$ is the variance in values ranging from 0 to 100 with increasing increments of 5.

### Goal 02: Does recombination change the topology of genomic samples?
<a href="url"><img src="https://github.com/alfonsomiguelcruz/pgc_btip_topone/blob/main/imgs/goal02_pipeline.png" height="50%" width="50%" ></a>

In this pipeline, virus sequence samples were taken from the [GISAID](https://gisaid.org/) database. The recombinant lineage and its parent lineage samples were downloaded, each stored in a `.fasta` file. The files were reorganized to produce three new FASTA files:

- Recombinant samples (recombinant lineage samples)
- Nonrecombinant samples (both parent lineage samples)
- Mixed samples (recombinant and parent lineage samples)

After organizing the sequences, all sequences in each file were aligned using the [Nextclade CLI](https://docs.nextstrain.org/projects/nextclade/en/stable/user/nextclade-cli/installation/standalone.html) against the [SARS-CoV-2 Wuhan reference genome](https://github.com/nextstrain/nextclade_data/blob/release/data/nextstrain/sars-cov-2/wuhan-hu-1/orfs/reference.fasta). The aligned sequences were processed further by removing a column of nucleotides in all sequences if at least one sequence contains an unknown nucleotide. This processing results in sequences containing the four nucleotides (`A`, `T`, `C`, `G`) and gaps (`-`) only.

Once the sequences have been processed, we continue a similar pipeline as shown in Goal 01, by producing the Hamming distance matrices from the sequences, and computing the birth and death times per homology group from the Hamming distance matrix.

After obtaining the birth and death times, they are used for statistical analysis by using the Kruskal-Wallis Test and Independent Mann-Whitney U Test at a 5% level of significance.

## Data
### `ms`-generated Sequences
The simulated samples were generated from `ms`. These sequences are in a biallelic format, where the characters in a sequence are `1` or `0` only.

### GISAID SARS-CoV-2 Sequences
The virus sequence samples, specifically SARS-CoV-2 samples, were taken from the GISAID database. Three recombinant lineages and their parent lineages were taken across different countries. The table below shows the lineages used and the countries whose samples were used:

| Recombinant Lineage | Parent Lineage 1 | Parent Lineage 2 | Countries
| --- | --- | --- | --- |
| XBC.1 | BA.2 | B.1.617.2 | China, Philippines, Singapore, South Korea, United States of America
| XBE | BA.5.2 | BE.4.1 | United Kingdom, United States of America
| XBZ | BA.5.2.1 | EF.1.3 | Austria, Denmark, Germany

## Execution
### Goal 01: Is topology robust to noise and sparse genomic samples?
Before generating the sequences, ensure that the software `ms` is inside this directory. To generate the simulated sequences, run the command on your terminal:

`./popgensims.sh`

to generate simulated sequences at differing values of mutation and recombination rates, and number of samples. The output of the simulated sequences are found at the [simseq](inputs/simseq/) folder. Once the sequences have been generated, the plots can now be created.

The script `goal_one_plots.py` uses simulated sequence data to produce the plots necessary to answer this question.

To execute the script, simply execute `python goal_one_plots.py` on your command line with any of the following arguments:

- `--get-one-plot`: gets one PDF file containing two plots: the variance and sparsity plots. Using this argument must also require the use of three additional arguments:
    
    - `-n`: simulated sample size $n$
    - `-t`: mutation rate $\theta$
    - `-r`: recombination rate $\rho$

- `--get-all-plots`: gets all PDF files containing the variance and sparsity plots, across every combination of possible simulated sample sizes, mutation rates, and recombination rates

- `--verbose`: Optional argument that outputs the logs the start and end of each step in the script.


Some sample commands are shown below:

- `python goal_one_plots.py --get-one-plot -n 100 -t 50 -r 72 --verbose`
- `python goal_one_plots.py --get-all-plots`

### Goal 02: Does recombination change the topology of genomic samples?
The script `goal_two_plots.py` uses the GISAID sequence data to produce the data required for the plots and for the statistical analysis.

To execute the script, simply execute `python goal_two_plots.py` on your command line with any of the following arguments:

- `--input`: Uses the sample Hamming distance matrices provided in the repository. The user can choose between `xbc.1`, `xbe`, and `xbz`.
- `--mats` : Uses the Hamming distance matrices provided by the user themselves.
- `--seqs` : Uses the FASTA files containing the aligned sequences provided by the user themselves.


## Results
<!-- Show Results of the Project in Paragraph form only -->
